openquake/hazardlib/gsim/gsim_table.py

```diff
@@ -0,0 +1,545 @@
+                    interpolator(rctx.mag).reshape(self.shape[0],
+                                                   self.shape[3]))
+
+            else:
+                # For spectral accelerations - need two step process
+                # Interpolate period
+                interpolator = interp1d(numpy.log10(self.periods),
+                                        self.sigma[stddev_type]["SA"],
+                                        axis=1)
+                period_table = interpolator(numpy.log10(imt.period))
+                mag_interpolator = interp1d(self.magnitudes,
+                                            period_table,
+                                            axis=1)
+                output_tables.append(mag_interpolator(rctx.mag))
+        return output_tables
+
+
+class GMPETable(GMPE):
+    """
+    Implements ground motion prediction equations in the form of a table from
+    which the expected ground motion intensity levels and standard deviations
+    are interpolated.
+
+    In a GMPE tables the expected ground motions for each of the IMTs over the
+    range of magnitudes and distances are stored in an hdf5 file on the path
+    specified by the user.
+
+    In this version of the GMPE the expected values are interpolated to the
+    required IMT, magnitude and distance in three stages.
+
+    i) Initially the correct IMT values are identified, interpolating in
+       log-T|log-IML space between neighbouring spectral periods.
+
+    ii) The IML values are then interpolated to the correct magnitude using
+        linear-M|log-IML space
+
+    iii) The IML values are then interpolated to the correct distance via
+         linear-D|linear-IML interpolation
+    """
+    DEFINED_FOR_TECTONIC_REGION_TYPE = ""
+
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set(())
+
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = ""
+
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set((const.StdDev.TOTAL,))
+
+    REQUIRES_SITES_PARAMETERS = set(())
+
+    REQUIRES_DISTANCES = set(())
+
+    REQUIRES_RUPTURE_PARAMETERS = {"mag"}
+
+    GMPE_TABLE = None
+
+    def __init__(self, gmpe_table=None):
+        """
+        Instantiate - either with a GMPE table or otherwise it will take
+        the GMPE table defined
+        """
+        if not self.GMPE_TABLE:
+            if gmpe_table:
+                self.GMPE_TABLE = gmpe_table
+            else:
+                raise IOError("GMPE Table Not Defined!")
+        super(GMPETable, self).__init__()
+        self.imls = None
+        self.stddevs = {}
+        self.m_w = None
+        self.distances = None
+        self.distance_type = None
+        self.amplification = None
+        self._run_setup()
+
+    def _run_setup(self):
+        """
+        Executes the preprocessing steps at the instantiation stage to read in
+        the tables from hdf5 and hold them in memory.
+        """
+        fle = h5py.File(self.GMPE_TABLE, "r")
+        self.distance_type = fle["Distances"].attrs["metric"]
```

Comment:
At this point trying to support multiple distance metrics and their respective scaling under different rupture conditions is a complex problem that would require much higher dimensionality than we can support here. If we had multiple distance metrics, which do we interpolate upon? I am inclined to wait for user feedback to see if we need this functionality.